docs/proposals/train_api_proposal.md

```diff
@@ -0,0 +1,154 @@
+
+```python
+training_client.create_job(
+   name=pytorchjob_name,
+   train_func=train_function,
+   num_worker_replicas=3, # How many PyTorch Workers will be created.
+)
+```
+
+The above method reduces the time taken but still the end user has to write the train_function method.
+
+In many cases like in LLM, training core code template is a standard and it doesn’t require any custom modification. For faster experimentation of such models, an easier api should exist to train/finetune models without writing custom Python code. HuggingFace Trainer class is a good example but it is not easy to set it up in a Kubernetes environment. Training operator can handle this complexity and provide a better abstracted layer to the user.
+
+**<h3>New Train API proposal:</h3>**
+
+```python
+From kubeflow.training import trainingClient
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+# Arguments related to the dataset provider including credentials
+dataset_args = datasetProviderClass()
+# Arguments related to the trainer code
+parameters = {key:value pairs}
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='provider://repo/model_name', 
+   dataset= 'provider://train_dataset_path',
+   eval_dataset = "provider://eval_dataset_path"
+   model_args, 
+   dataset_args, 
+   parameters
+)
+```
+
+Example: 
+
+```python
+@dataclass
+class HuggingFace:
+    access_token = field()
+    peft_config = field()
+    transformerClass = field()
+
+@dataclass
+class S3:
+    access_token = field()
+    region = field()
+ 
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='hf://openchat/openchat_3.5', 
+   dataset= 's3://doc-example-bucket1/train_dataset',
+   eval_dataset = "s3://doc-example-bucket1/eval_dataset"
+   HuggingFace(access_token = "hf_..." , peft_config = {lora_alpha: 16}, transformerClass="Trainer"), 
+   s3(access_token = "s3 access token", region="us-west-2"), 
+   {learning_rate=0.1}
+)
+```
+
+The new proposed API takes following arguments 
+
+1. System parameters - Number of nodes, number of procs per node(GPUs per node).
+2. Model parameters - Model provider and repository details.
+3. Dataset parameters - Dataset provider and dataset details.
+4. Training parameters - Training specific parameters like learning rate etc.
+
+**<h3>Implementation</h3>**
+
+1. Setup **init** **containers** that download the model and dataset to a PVC. Based on the specified model provider, corresponding training utility functions will be used. Eg: For Huggingface provider, Huggingface trainer can be used. For this **get_pytorchjob_template** function in the sdk needs to be changed to add init containers spec.. Inorder to download models and data sets, we need to support different providers like kaggle, hugging face, s3 or git lfs. The data can be stored in a shared volume between the init container and the main container.
```

Comment:
Do we need to maintain separate images for Storage Initialiser and Actual Trainer ?
I think, we need to discuss how to effectively download models/datasets for multiple workers in PyTorchJobs

Reply:
I think separation will help in keeping the actual trainer image simple.
Currently KServe comes with a default storage initializer image (https://github.com/kserve/kserve/blob/61c9bd334ae9766ffd1f2bf020764bf453cab54c/python/kserve/kserve/storage/storage.py#L59) for downloading models. It supports common storage protocols like S3, GCS, Azure. I can reuse this.

Reply:
Can we use unify Storage Initializer image for model and dataset download ?